File: src/main/java/com/architecture_map/belarus/entity/ArchitecturalStyle.java
package com.architecture_map.belarus.entity;

import com.architecture_map.belarus.entity.image.Image;
import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.RequiredArgsConstructor;
import lombok.Setter;
import lombok.ToString;

import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.JoinColumn;
import javax.persistence.JoinTable;
import javax.persistence.ManyToMany;
import javax.persistence.OneToOne;

import java.util.HashSet;
import java.util.Set;

@Entity
@Getter
@Setter(AccessLevel.PUBLIC)
@RequiredArgsConstructor
@ToString
public class ArchitecturalStyle {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Integer id;

    private String name;
    private String description;

    @ManyToMany
    @JoinTable(
            name = "architectural_style_architectural_attribute",
            joinColumns = {@JoinColumn(name = "architectural_style_id")},
            inverseJoinColumns = {@JoinColumn(name = "architectural_attribute_id")}
    )
    private Set<ArchitecturalAttribute> attributes = new HashSet<>();

    @JsonIgnoreProperties({"construction"})
    @OneToOne
    @JoinColumn(name = "demonstrative_image_id", referencedColumnName = "id")
    private Image demonstrativeImage;

    private String shortDescription;
    private String yearsActive;
}

File: src/main/kotlin/by/architecturemap/belarus/entity/ArchitecturalStyle.kt
package by.architecturemap.belarus.entity

import com.fasterxml.jackson.annotation.JsonIgnoreProperties
import jakarta.persistence.CascadeType
import jakarta.persistence.Entity
import jakarta.persistence.JoinColumn
import jakarta.persistence.JoinTable
import jakarta.persistence.ManyToMany
import jakarta.persistence.OneToOne
import jakarta.validation.constraints.NotBlank

@Entity
data class ArchitecturalStyle(

    @NotBlank(message = "Name may not be blank")
    var name: String,

    var description: String? = null,

    @ManyToMany(cascade = [CascadeType.PERSIST, CascadeType.REMOVE])
    @JoinTable(
        name = "architectural_style_architectural_attribute",
        joinColumns = [JoinColumn(name = "architectural_style_id")],
        inverseJoinColumns = [JoinColumn(name = "architectural_attribute_id")]
    )
    var attributes: List<ArchitecturalAttribute>? = listOf(),

    @JsonIgnoreProperties("construction")
    @OneToOne(cascade = [CascadeType.PERSIST, CascadeType.REMOVE])
    @JoinColumn(name = "demonstrative_image_id", referencedColumnName = "id")
    var demonstrativeImage: Image? = null,

    var shortDescription: String? = null,
    var yearsActive: String? = null

) : BaseEntity()
